internal/venice-common/src/main/java/com/linkedin/venice/writer/VeniceWriterFactory.java

```diff
@@ -61,35 +61,47 @@ public VeniceWriterFactory(
   }
 
   public <K, V, U> VeniceWriter<K, V, U> createVeniceWriter(VeniceWriterOptions options) {
-    VeniceProperties props = options.isProducerCompressionEnabled()
-        ? venicePropertiesLazy.get()
-        : venicePropertiesWithCompressionDisabledLazy.get();
+    VeniceProperties props = getVeniceProperties(options);
+    Supplier<PubSubProducerAdapter> producerAdapterSupplier =
+        () -> producerAdapterFactory.create(props, options.getTopicName(), options.getBrokerAddress());
+    PubSubProducerAdapter producerAdapter = buildPubSubProducerAdapter(options, producerAdapterSupplier);
+    return new VeniceWriter<>(options, props, producerAdapter);
+  }
 
+  public <K, V, U> ComplexVeniceWriter<K, V, U> createComplexVeniceWriter(VeniceWriterOptions options) {
```

Comment:
Shall we enforce that `ComplexVeniceWriter` will always use `ComplexVenicePartitioner`? otherwise, the logic is a little confusing to me.